### PR TITLE
Fix error in deliver files

### PR DESCRIPTION
**Pull request by diitaz93**
## Description
The DeliverFiles model currently does not allow for analysis sample files to be empty. This PR allows that, adds debug logging, remove error raising and instead prints to stdout when there are no files to deliver

Arose form problems delivering a microsalt fastq-qc-analysis case goldenskylark

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b fix-delivery-error-handling -a
    ```

### How to test

- [x] Do `cg -l DEBUG deliver case -c goldenskylark --dry-run`
- [x] Do `cg -l DEBUG -c config-hk-prod.yaml deliver case -c goldenskylark --dry-run`

### Expected test outcome

- [x] Check that when there are no files to deliver, the command prints that in stdout
- [x] Check that if there are files to deliver, they are linked to the customer's folder

## Review

- [x] Tests executed by SD
- [x] "Merge and deploy" approved by VJ
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/services/deliver_files/deliver_files_service/deliver_files_service.py [hunk 1]
```diff
@@ -89,6 +89,7 @@ def deliver_files_for_ticket(
             )
 
     def _start_rsync_job(self, case: Case, dry_run: bool, folders_to_deliver: set[Path]) -> int:
+        LOG.debug(f"[RSYNC] Starting rsync job for case {case.internal_id}")
```
Comment (diitaz93) [inline cg/services/deliver_files/deliver_files_service/deliver_files_service.py:92]: Do you like this logging? @Vince-janv I add the service at the beginning of the error message

File: cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py [hunk 2]
```diff
@@ -61,6 +63,9 @@ def _get_sample_files_from_case_bundle(
         sample_files: list[File] = self.hk_api.get_files_from_latest_version_containing_tags(
             bundle_name=case_id, tags=sample_tags_with_sample_id
         )
+        if not sample_files:
+            LOG.warning(f"No files found for sample {sample_id} in case {case_id}")
```
Comment (diitaz93) [inline cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py:67]: Maybe this one is not that necessary

**Comment by diitaz93** [comment]
## Tests on stage
```shell
$ cg -l DEBUG deliver case -c goldenskylark --dry-run
Running CG deliver
Instantiating delivery service factory
Instantiating status db
Instantiating housekeeper api
Initializing Store
Instantiating trailblazer api
Called undefined __fields__ on HousekeeperAPI, please wrap
Instantiating delivery rsync service
[FETCH SERVICE] Fetching raw data files for case: goldenskylark
Fetch latest version from bundle sample_id
Bundle: sample_id not found in Housekeeper
...
....

[FETCH SERVICE] Fetching analysis files for case: goldenskylark
Fetch latest version from bundle goldenskylark
Bundle: goldenskylark not found in Housekeeper
...
...
No files to deliver for case goldenskylark
[js.diazboada@hasta:~] [S_base] $
```

**Comment by diitaz93** [comment]
## Test on stage: Found only raw-data files
```shell
$ cg -l DEBUG -c cg-stage-hk-prod.yaml deliver case -c goldenskylark --dry-run
Running CG deliver
Instantiating delivery service factory
Instantiating status db
Instantiating housekeeper api
Initializing Store
Instantiating trailblazer api
Called undefined __fields__ on HousekeeperAPI, please wrap
Instantiating delivery rsync service
[FETCH SERVICE] Fetching raw data files for case: goldenskylark
Fetch latest version from bundle sample_id
Fetching files from version 201342
...
...
[FETCH SERVICE] Fetching analysis files for case: goldenskylark
Fetch latest version from bundle goldenskylark
Fetching files from version 201500
Fetch latest version from bundle goldenskylark
Fetching files from version 201500
No files found for sample sample_id in case goldenskylark
...
...
[MOVE SERVICE] Creating ticket inbox folder: /home/proj/stage/customers/cust056/inbox/1000309
[MOVE SERVICE] Creating hard links for delivery files in: /home/proj/stage/customers/cust056/inbox/1000309
/home/proj/stage/..._1.html already exists. Overwriting with /home/proj/production/housekeeper-bundles/goldenskylark/2024-09-23/1000309_QC_1.html
Linked /home/proj/production/housekeeper-bundles/goldenskylark/2024-09-23/1000309_QC_1.html to /home/proj/stage/...1000309_QC_1.html
...
....
[MOVE SERVICE] Replacing file paths with inbox dir path: /home/proj/stage/customers/cust056/inbox/1000309
[FORMAT SERVICE] Formatting files for delivery
Concatenated forward reads to /home/proj/stage/...1.fastq.gz
Concatenated reverse reads to /home/proj/stage/..._2.fastq.gz
...
...
[RSYNC] Starting rsync job for case goldenskylark
Setting log dir to: /home/proj/stage/rsync/goldenskylark_240925_12_30_43_140855
Creating folder: /home/proj/stage/rsync/goldenskylark_240925_12_30_43_140855
Would have created path /home/proj/stage/rsync/goldenskylark_240925_12_30_43_140855, but this is a dry run
Initialising Process with binary: sbatch
Use base call ['sbatch']
SlurmAPI: Set dry run to True
Write sbatch content to path /home/proj/stage/rsync/goldenskylark_240925_12_30_43_140855/goldenskylark_rsync.sh:
#! /bin/bash
#SBATCH --job-name=goldenskylark_rsync
#SBATCH --account=development
#SBATCH --ntasks=1
#SBATCH --mem=1G
#SBATCH --error=
#SBATCH --output=
#SBATCH --mail-type=FAIL
#SBATCH --mail-user=clinical-logwatch@scilifelab.se
#SBATCH --time=24:00:00
#SBATCH --qos=low
#SBATCH --exclude=gpu-compute-0-[0-1],cg-dragen

...
...

Submit sbatch /home/proj/stage/rsync/goldenskylark_240925_12_30_43_140855/goldenskylark_rsync.sh
Running command sbatch /home/proj/stage/rsync/goldenskylark_240925_12_30_43_140855/goldenskylark_rsync.sh
Dry run: process call will not be executed!!

Could not get slurm job number
Dry-run: it would write slurm jobs to /home/proj/stage/rsync/goldenskylark_240925_12_30_43_140855/slurm_job_ids.yaml
Would have added the analysis for case goldenskylark to Trailblazer
[js.diazboada@hasta:~] [S_base] $
```

File: cg/services/deliver_files/delivery_file_formatter_service/delivery_file_formatter.py [hunk 3]
```diff
@@ -42,6 +46,7 @@ def __init__(
 
     def format_files(self, delivery_files: DeliveryFiles) -> FormattedFiles:
         """Format the files to be delivered and return the formatted files in the generic format."""
+        LOG.debug(f"[FORMAT SERVICE] Formatting files for delivery")
```
Comment (diitaz93) [inline cg/services/deliver_files/delivery_file_formatter_service/delivery_file_formatter.py:49]:
```suggestion
        LOG.debug("[FORMAT SERVICE] Formatting files for delivery")
```

File: cg/services/deliver_files/deliver_files_service/deliver_files_service.py [hunk 4]
```diff
@@ -105,6 +112,7 @@ def _add_trailblazer_tracking(self, case: Case, job_id: int, dry_run: bool) -> N
         if dry_run:
             LOG.info(f"Would have added the analysis for case {case.internal_id} to Trailblazer")
         else:
+            LOG.debug(f"[TB SERVICE] Adding analysis for case {case.internal_id} to Trailblazer")
```
Comment (diitaz93) [inline cg/services/deliver_files/deliver_files_service/deliver_files_service.py:115]: lines 126&127 seem to be doing the same

File: cg/services/deliver_files/deliver_files_service/deliver_files_service.py [hunk 5]
```diff
@@ -56,14 +55,16 @@ def __init__(
...
-    @handle_delivery_errors
     def deliver_files_for_case(
         self, case: Case, delivery_base_path: Path, dry_run: bool = False
     ) -> None:
         """Deliver the files for a case to the customer folder."""
         delivery_files: DeliveryFiles = self.file_manager.get_files_to_deliver(
             case_id=case.internal_id
         )
+        if not self._are_files_to_deliver(delivery_files):
+            LOG.error(f"No files to deliver for case {case.internal_id}")
```
Comment (Vince-janv) [inline cg/services/deliver_files/deliver_files_service/deliver_files_service.py:66]:
```suggestion
            LOG.warning(f"No files to deliver for case {case.internal_id}")
```

File: cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py [hunk 6]
```diff
@@ -61,6 +63,9 @@ def _get_sample_files_from_case_bundle(
...
+            return None
```
Comment (Vince-janv) [inline cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py:68]: Remove to instead return empy list

File: cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py [hunk 7]
```diff
@@ -77,10 +82,10 @@ def _get_analysis_sample_delivery_files(self, case: Case) -> list[SampleFile] |
         sample_ids: list[str] = case.sample_ids
         delivery_files: list[SampleFile] = []
         for sample_id in sample_ids:
-            sample_delivery_files: list[SampleFile] = self._get_sample_files_from_case_bundle(
+            sample_files: list[SampleFile] | None = self._get_sample_files_from_case_bundle(
                 case_id=case.internal_id, sample_id=sample_id, workflow=case.data_analysis
             )
-            delivery_files.extend(sample_delivery_files) if sample_delivery_files else None
+            delivery_files.extend(sample_files) if sample_files else None
```
Comment (Vince-janv) [inline cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py:88]: on line 92 remove the None return type

**Review by Vince-janv — APPROVED** [review]
Would be nice to skip the `| None` everywhere and just work with lists. Otherwise it looks good!

File: cg/services/deliver_files/delivery_file_fetcher_service/models.py [hunk 8]
```diff
@@ -24,6 +24,4 @@ class SampleFile(BaseModel):
 class DeliveryFiles(BaseModel):
     delivery_data: DeliveryMetaData
     case_files: list[CaseFile] | None = None
-    sample_files: list[SampleFile] = Field(
-        ..., min_length=1, description="At least one sample file is required for delivery."
-    )
+    sample_files: list[SampleFile] | None = None
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/services/deliver_files/delivery_file_fetcher_service/models.py:27]: fixed

File: cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py (hunk 7 above)
Comment (diitaz93) [reply cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py:88]: fixed

File: cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py (hunk 6 above)
Comment (diitaz93) [reply cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py:68]: fixed

File: cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py (hunk 2 above)
Comment (diitaz93) [reply cg/services/deliver_files/delivery_file_fetcher_service/analysis_delivery_file_fetcher.py:67]: removing

**Pull request merged**